routers/web/web.go

```diff
@@ -797,12 +798,12 @@ func RegisterRoutes(m *web.Route) {
 				m.Combo("/_upload/*", repo.MustBeAbleToUpload).
 					Get(repo.UploadFile).
 					Post(bindIgnErr(forms.UploadRepoFileForm{}), repo.UploadFilePost)
-			}, context.RepoRefByType(context.RepoRefBranch), repo.MustBeEditable)
+			}, context.RepoRefByType(context.RepoRefBranch), canEnableEditor, repo.MustBeEditable)
 			m.Group("", func() {
 				m.Post("/upload-file", repo.UploadFileToServer)
 				m.Post("/upload-remove", bindIgnErr(forms.RemoveUploadFileForm{}), repo.RemoveUploadFileFromServer)
 			}, context.RepoRef(), repo.MustBeEditable, repo.MustBeAbleToUpload)
-		}, context.RepoMustNotBeArchived(), reqRepoCodeWriter, repo.MustBeNotEmpty)
+		}, context.RepoMustNotBeArchived(), repo.MustBeNotEmpty)
```

Comment:
`reqRepoCodeWriter` also protects other handler like `m.Post("/upload-file",` , why is it deleted?

Reply:
Because all the protected endpoints should be usable if you have branch write access (you are also able to upload files)

Reply:
But the new `canEnableEditor` was written in a inner level, is it correct?

![image](https://user-images.githubusercontent.com/2114189/147545193-56e124e2-4429-4226-960b-72bbbf1fe132.png)

Reply:
Yes, it needs to run after `context.RepoRefBranch` to get the branch name

Reply:
Would `m.Post("/upload-file"` lose the access protection and then everyone can write it?

Reply:
Maybe we need to replace `reqRepoCodeWriter` as another one or change `reqRepoCodeWriter`.